File: android/src/main/java/expo/modules/alarm/ExpoAlarmModule.kt
package expo.modules.alarm

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.util.*

class ExpoAlarmModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()
  
  private val alarmManager: AlarmManager
    get() = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
  
  private val sharedPreferences: SharedPreferences
    get() = context.getSharedPreferences("ExpoAlarmModule", Context.MODE_PRIVATE)
  
  private val gson = Gson()

  override fun definition() = ModuleDefinition {
    Name("ExpoAlarm")

    Events("alarmTriggered", "alarmDismissed")

    Function("isSupported") {
      true // Android always supports alarms
    }

    AsyncFunction("requestPermissionsAsync") { promise: Promise ->
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        // Android 12+ requires SCHEDULE_EXACT_ALARM permission
        val hasPermission = alarmManager.canScheduleExactAlarms()
        promise.resolve(mapOf(
          "granted" to hasPermission,
          "canAskAgain" to !hasPermission
        ))
      } else {
        promise.resolve(mapOf(
          "granted" to true,
          "canAskAgain" to false
        ))
      }
    }

    AsyncFunction("getPermissionsAsync") { promise: Promise ->
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        val hasPermission = alarmManager.canScheduleExactAlarms()
        promise.resolve(mapOf(
          "granted" to hasPermission,
          "canAskAgain" to !hasPermission
        ))
      } else {
        promise.resolve(mapOf(
          "granted" to true,
          "canAskAgain" to false
        ))
      }
    }

    AsyncFunction("scheduleAlarmAsync") { alarmData: Map<String, Any>, promise: Promise ->
      try {
        val identifier = alarmData["identifier"] as String
        val title = alarmData["title"] as String
        val body = alarmData["body"] as? String
        val dateMillis = (alarmData["date"] as Double).toLong()
        val repeating = alarmData["repeating"] as? Boolean ?: false
        val repeatInterval = (alarmData["repeatInterval"] as? Double)?.toLong()
        val sound = alarmData["sound"] as? String

        // Cancel existing alarm with same identifier
        cancelAlarmInternal(identifier)

        // Create alarm intent
        val intent = Intent(context, AlarmReceiver::class.java).apply {
          putExtra("identifier", identifier)
          putExtra("title", title)
          putExtra("body", body)
          putExtra("repeating", repeating)
          putExtra("repeatInterval", repeatInterval)
          putExtra("sound", sound)
        }

        val requestCode = identifier.hashCode()
        val pendingIntent = PendingIntent.getBroadcast(
          context,
          requestCode,
          intent,
          PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        // Schedule alarm
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && !alarmManager.canScheduleExactAlarms()) {
          throw Exception("Exact alarm permission not granted")
        }

        if (repeating && repeatInterval != null) {
          alarmManager.setRepeating(
            AlarmManager.RTC_WAKEUP,
            dateMillis,
            repeatInterval,
            pendingIntent
          )
        } else {
          alarmManager.setExactAndAllowWhileIdle(
            AlarmManager.RTC_WAKEUP,
            dateMillis,
            pendingIntent
          )
        }

        // Store alarm info
        val alarmInfo = mapOf(
          "identifier" to identifier,
          "title" to title,
          "body" to body,
          "date" to dateMillis,
          "repeating" to repeating,
          "repeatInterval" to repeatInterval,
          "sound" to sound,
          "enabled" to true
        )
        
        saveAlarmInfo(identifier, alarmInfo)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("ERR_ALARM_SCHEDULE", e.message, e)
      }
    }

    AsyncFunction("cancelAlarmAsync") { identifier: String, promise: Promise ->
      try {
        cancelAlarmInternal(identifier)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("ERR_ALARM_CANCEL", e.message, e)
      }
    }

    AsyncFunction("cancelAllAlarmsAsync") { promise: Promise ->
      try {
        val alarms = getAllStoredAlarms()
        for (alarmId in alarms.keys) {
          cancelAlarmInternal(alarmId)
        }
        clearAllAlarms()
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("ERR_ALARM_CANCEL_ALL", e.message, e)
      }
    }

    AsyncFunction("getAllAlarmsAsync") { promise: Promise ->
      try {
        val alarms = getAllStoredAlarms()
        val alarmList = alarms.values.toList()
        promise.resolve(alarmList)
      } catch (e: Exception) {
        promise.reject("ERR_ALARM_GET_ALL", e.message, e)
      }
    }

    AsyncFunction("getAlarmAsync") { identifier: String, promise: Promise ->
      try {
        val alarmInfo = getStoredAlarmInfo(identifier)
        promise.resolve(alarmInfo)
      } catch (e: Exception) {
        promise.reject("ERR_ALARM_GET", e.message, e)
      }
    }

    AsyncFunction("hasAlarmAsync") { identifier: String, promise: Promise ->
      try {
        val hasAlarm = getStoredAlarmInfo(identifier) != null
        promise.resolve(hasAlarm)
      } catch (e: Exception) {
        promise.reject("ERR_ALARM_HAS", e.message, e)
      }
    }
  }

  private fun cancelAlarmInternal(identifier: String) {
    val intent = Intent(context, AlarmReceiver::class.java)
    val requestCode = identifier.hashCode()
    val pendingIntent = PendingIntent.getBroadcast(
      context,
      requestCode,
      intent,
      PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
    )
    
    alarmManager.cancel(pendingIntent)
    removeStoredAlarmInfo(identifier)
  }

  private fun saveAlarmInfo(identifier: String, alarmInfo: Map<String, Any?>) {
    val editor = sharedPreferences.edit()
    val json = gson.toJson(alarmInfo)
    editor.putString("alarm_$identifier", json)
    editor.apply()
  }

  private fun getStoredAlarmInfo(identifier: String): Map<String, Any?>? {
    val json = sharedPreferences.getString("alarm_$identifier", null)
    return if (json != null) {
      val type = object : TypeToken<Map<String, Any?>>() {}.type
      gson.fromJson(json, type)
    } else {
      null
    }
  }

  private fun removeStoredAlarmInfo(identifier: String) {
    val editor = sharedPreferences.edit()
    editor.remove("alarm_$identifier")
    editor.apply()
  }

  private fun getAllStoredAlarms(): Map<String, Map<String, Any?>> {
    val allEntries = sharedPreferences.all
    val alarms = mutableMapOf<String, Map<String, Any?>>()
    
    for ((key, value) in allEntries) {
      if (key.startsWith("alarm_") && value is String) {
        val identifier = key.removePrefix("alarm_")
        val type = object : TypeToken<Map<String, Any?>>() {}.type
        val alarmInfo: Map<String, Any?> = gson.fromJson(value, type)
        alarms[identifier] = alarmInfo
      }
    }
    
    return alarms
  }

  private fun clearAllAlarms() {
    val editor = sharedPreferences.edit()
    val allEntries = sharedPreferences.all
    
    for (key in allEntries.keys) {
      if (key.startsWith("alarm_")) {
        editor.remove(key)
      }
    }
    
    editor.apply()
  }
}

File: ios/ExpoAlarmModule.swift
import ExpoModulesCore
import UserNotifications
import Foundation

@available(iOS 16.0, *)
import AlarmKit

public class ExpoAlarmModule: Module {
  private var storedAlarms: [String: [String: Any]] = [:]
  private let userDefaults = UserDefaults.standard
  private let alarmsKey = "ExpoAlarmModule_alarms"
  
  public override func definition() -> ModuleDefinition {
    Name("ExpoAlarm")

    Events("alarmTriggered", "alarmDismissed")

    Function("isSupported") {
      if #available(iOS 16.0, *) {
        return true // AlarmKit available
      } else {
        return true // fallback to notifications
      }
    }

    AsyncFunction("requestPermissionsAsync") { (promise: Promise) in
      UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) { granted, error in
        if let error = error {
          promise.reject("ERR_PERMISSIONS", error.localizedDescription, error)
          return
        }
        
        promise.resolve([
          "granted": granted,
          "canAskAgain": !granted
        ])
      }
    }

    AsyncFunction("getPermissionsAsync") { (promise: Promise) in
      UNUserNotificationCenter.current().getNotificationSettings { settings in
        let granted = settings.authorizationStatus == .authorized
        promise.resolve([
          "granted": granted,
          "canAskAgain": settings.authorizationStatus != .denied
        ])
      }
    }

    AsyncFunction("scheduleAlarmAsync") { (alarmData: [String: Any], promise: Promise) in
      do {
        guard let identifier = alarmData["identifier"] as? String,
              let title = alarmData["title"] as? String,
              let dateMillis = alarmData["date"] as? Double else {
          promise.reject("ERR_INVALID_ARGS", "Missing required fields", nil)
          return
        }
        
        let body = alarmData["body"] as? String
        let repeating = alarmData["repeating"] as? Bool ?? false
        let repeatInterval = alarmData["repeatInterval"] as? Double
        let sound = alarmData["sound"] as? String
        
        let date = Date(timeIntervalSince1970: dateMillis / 1000)
        
        // Cancel existing alarm with same identifier
        await self.cancelAlarmInternal(identifier: identifier)
        
        if #available(iOS 16.0, *) {
          // Use AlarmKit for iOS 16+
          try await self.scheduleWithAlarmKit(
            identifier: identifier,
            title: title,
            body: body,
            date: date,
            repeating: repeating,
            repeatInterval: repeatInterval,
            sound: sound
          )
        } else {
          // Use UserNotifications for older iOS versions
          try await self.scheduleWithNotifications(
            identifier: identifier,
            title: title,
            body: body,
            date: date,
            repeating: repeating,
            repeatInterval: repeatInterval,
            sound: sound
          )
        }
        
        // Store alarm info
        let alarmInfo: [String: Any] = [
          "identifier": identifier,
          "title": title,
          "body": body ?? NSNull(),
          "date": dateMillis,
          "repeating": repeating,
          "repeatInterval": repeatInterval ?? NSNull(),
          "sound": sound ?? NSNull(),
          "enabled": true
        ]
        
        self.saveAlarmInfo(identifier: identifier, alarmInfo: alarmInfo)
        promise.resolve(nil)
        
      } catch {
        promise.reject("ERR_ALARM_SCHEDULE", error.localizedDescription, error)
      }
    }

    AsyncFunction("cancelAlarmAsync") { (identifier: String, promise: Promise) in
      do {
        await self.cancelAlarmInternal(identifier: identifier)
        promise.resolve(nil)
      } catch {
        promise.reject("ERR_ALARM_CANCEL", error.localizedDescription, error)
      }
    }

    AsyncFunction("cancelAllAlarmsAsync") { (promise: Promise) in
      do {
        let alarms = self.getAllStoredAlarms()
        for alarmId in alarms.keys {
          await self.cancelAlarmInternal(identifier: alarmId)
        }
        self.clearAllAlarms()
        promise.resolve(nil)
      } catch {
        promise.reject("ERR_ALARM_CANCEL_ALL", error.localizedDescription, error)
      }
    }

    AsyncFunction("getAllAlarmsAsync") { (promise: Promise) in
      let alarms = self.getAllStoredAlarms()
      let alarmList = Array(alarms.values)
      promise.resolve(alarmList)
    }

    AsyncFunction("getAlarmAsync") { (identifier: String, promise: Promise) in
      let alarmInfo = self.getStoredAlarmInfo(identifier: identifier)
      promise.resolve(alarmInfo)
    }

    AsyncFunction("hasAlarmAsync") { (identifier: String, promise: Promise) in
      let hasAlarm = self.getStoredAlarmInfo(identifier: identifier) != nil
      promise.resolve(hasAlarm)
    }
  }
  
  @available(iOS 16.0, *)
  private func scheduleWithAlarmKit(identifier: String, title: String, body: String?, date: Date, repeating: Bool, repeatInterval: Double?, sound: String?) async throws {
    // AlarmKit implementation would go here
    // For now, fall back to notifications since AlarmKit has limited availability
    try await scheduleWithNotifications(identifier: identifier, title: title, body: body, date: date, repeating: repeating, repeatInterval: repeatInterval, sound: sound)
  }
  
  private func scheduleWithNotifications(identifier: String, title: String, body: String?, date: Date, repeating: Bool, repeatInterval: Double?, sound: String?) async throws {
    let content = UNMutableNotificationContent()
    content.title = title
    content.body = body ?? ""
    content.sound = UNNotificationSound.default
    
    let calendar = Calendar.current
    let components = calendar.dateComponents([.year, .month, .day, .hour, .minute, .second], from: date)
    
    let trigger: UNNotificationTrigger
    if repeating, let interval = repeatInterval {
      trigger = UNTimeIntervalNotificationTrigger(timeInterval: interval / 1000, repeats: true)
    } else {
      trigger = UNCalendarNotificationTrigger(dateMatching: components, repeats: false)
    }
    
    let request = UNNotificationRequest(identifier: identifier, content: content, trigger: trigger)
    
    try await UNUserNotificationCenter.current().add(request)
  }
  
  private func cancelAlarmInternal(identifier: String) async {
    // Cancel notification
    UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: [identifier])
    
    // Remove from storage
    removeStoredAlarmInfo(identifier: identifier)
  }
  
  private func saveAlarmInfo(identifier: String, alarmInfo: [String: Any]) {
    var allAlarms = getAllStoredAlarms()
    allAlarms[identifier] = alarmInfo
    
    if let data = try? JSONSerialization.data(withJSONObject: allAlarms, options: []) {
      userDefaults.set(data, forKey: alarmsKey)
    }
  }
  
  private func getStoredAlarmInfo(identifier: String) -> [String: Any]? {
    let allAlarms = getAllStoredAlarms()
    return allAlarms[identifier]
  }
  
  private func removeStoredAlarmInfo(identifier: String) {
    var allAlarms = getAllStoredAlarms()
    allAlarms.removeValue(forKey: identifier)
    
    if let data = try? JSONSerialization.data(withJSONObject: allAlarms, options: []) {
      userDefaults.set(data, forKey: alarmsKey)
    }
  }
  
  private func getAllStoredAlarms() -> [String: [String: Any]] {
    guard let data = userDefaults.data(forKey: alarmsKey),
          let alarms = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: [String: Any]] else {
      return [:]
    }
    return alarms
  }
  
  private func clearAllAlarms() {
    userDefaults.removeObject(forKey: alarmsKey)
  }
}
